### PR TITLE
Fix postprocess when value is a dict with only one key

### LINKED ISSUE #2191: Returning a dict from a function doesn't work as expected when the dict only has one key
### Describe the bug

When you return a dictionary from a function, the component may not be updated as expected if the dictionary has one key.

This is because the dict will get converted into a tuple of lists as opposed to a list with one element.

The following code handles dict outputs and you can see that it converts the predictions into a list

```python
        if type(predictions) is dict and len(predictions) > 0:
            keys_are_blocks = [isinstance(key, Block) for key in predictions.keys()]
            if all(keys_are_blocks):
                reordered_predictions = [skip() for _ in dependency["outputs"]]
                for component, value in predictions.items():
                    if component._id not in dependency["outputs"]:
                        return ValueError(
                            f"Returned component {component} not specified as output of function."
                        )
                    output_index = dependency["outputs"].index(component._id)
                    reordered_predictions[output_index] = value
                predictions = reordered_predictions
```
Then we convert that to a tuple:

```python
        if len(dependency["outputs"]) == 1:
            predictions = (predictions,)
```

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Reproduction

```python
import gradio as gr
import random
import string

with gr.Blocks() as demo:
    num = gr.Number()
    text = gr.Textbox()
    update = gr.Button(value="update")

    def update_values():
        return {text: random.choice(string.ascii_lowercase)}

    update.click(update_values, inputs=None, outputs=[text])

demo.launch()
```    

### Screenshot

You can see that the text displays `['v']` as opposed to just the character "v"

image

### Logs

```shell
-
```

### System Info

```shell
3.2
```

### Severity

annoying

### PR DESCRIPTION
**Pull request by freddyaboulton**
# Description
Fixes #2191 

# Checklist:

- [x] I have performed a self-review of my own code
- [x] My code follows the style guidelines of this project
- [x] I have commented my code in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: test/test_blocks.py [hunk 1]
```diff
@@ -279,6 +279,25 @@ def test_blocks_does_not_replace_keyword_literal():
...
 
+def test_blocks_returns_correct_output_dict_single_key():
+
+    with gr.Blocks() as demo:
+        num = gr.Number()
+        num2 = gr.Number()
+        update = gr.Button(value="update")
+
+        def update_values():
+            return {output: output.update(value=42)}
```
Comment (abidlabs) [inline test/test_blocks.py:290]: I think this should be as follows:
```suggestion
        def update_values(inp):
            return {num2: gr.update(value=42)}
```

File: test/test_blocks.py [hunk 2]
```diff
@@ -279,6 +279,25 @@ def test_blocks_does_not_replace_keyword_literal():
...
+
+        update.click(update_values, inputs=[num], outputs=[num2])
+
+    output = demo.postprocess_data(0, {num2: num2.update(value=42)}, state=None)
```
Comment (abidlabs) [inline test/test_blocks.py:294]:
```suggestion
    output = demo.postprocess_data(0, {num2: gr.Number.update(value=42)}, state=None)
```

File: gradio/blocks.py [hunk 3]
```diff
@@ -671,6 +671,11 @@ def postprocess_data(self, fn_index, predictions, state):
                         )
                     output_index = dependency["outputs"].index(component._id)
                     reordered_predictions[output_index] = value
+                # Unwrap list if only one output component to prevent
+                # wrapping the list into a tuple in the if len(reordered_predictions) == 1
+                # check
```
Comment (abidlabs) [inline gradio/blocks.py:676]: I would remove this long inline comment
```suggestion
```

File: gradio/blocks.py [hunk 4]
```diff
@@ -671,6 +671,11 @@ def postprocess_data(self, fn_index, predictions, state):
...
+                if len(reordered_predictions) == 1:
+                    reordered_predictions = reordered_predictions[0]
```
Comment (abidlabs) [inline gradio/blocks.py:678]: We've used this utils function in many other places in the code
```suggestion
                    reordered_predictions = utils.resolve_singleton(reordered_predictions)
```

File: gradio/blocks.py [hunk 5]
```diff
@@ -671,6 +671,11 @@ def postprocess_data(self, fn_index, predictions, state):
...
                 predictions = reordered_predictions
```
Comment (abidlabs) [inline gradio/blocks.py:679]: We've used this utils function in many other places in the code
```suggestion
                    predictions = utils.resolve_singleton(reordered_predictions)
```

**Comment by abidlabs** [comment]
LGTM, some small suggestions above. Also noticed that returning an empty dictionary `{}` as the output of a function results in an error. Instead, it should have no effect.

We could replace line 662: `if type(predictions) is dict and len(predictions) > 0:`, with the following:

```py
        if type(predictions) is dict:
             if len(predictions) == 0:
                 return []
```

to fix that

**Review by abidlabs — APPROVED** [review]

File: test/test_blocks.py (hunk 1 above)
Comment (freddyaboulton) [reply test/test_blocks.py:290]: Yep you're right it's technically a class method.  Both work and I sometimes inadvertently use it as an instance method cause of the faster autocomplete. Will make the change!

**Comment by freddyaboulton** [comment]
@abidlabs If we make that change, then you can't return an empty dictionary for JSON output. I'll keep that part of the code as is for now.

**Pull request merged**